File: Sources/KakaoSapiens/Services/SyncPairingPayload.swift
import Foundation

/// The thing a joining device scans.
///
/// It carries the address, the account being joined, the session, and the
/// `pairing_secret` — and nothing else. No device token, no master key, no
/// claim secret, no recovery phrase. Whoever holds this payload can attempt to
/// join, so it is shown on screen and never written to a URL, the clipboard, or
/// a log: a URL would make a scanner open a browser, leaving the secret in a
/// history somewhere.
public struct SyncPairingPayload: Equatable {
    public let baseURL: URL
    public let accountID: String
    public let sessionID: String
    public let pairingSecret: Data

    public enum PayloadError: Error, Equatable {
        case malformed
        case notCanonical
        case unsupportedVersion
    }

    private static let magic = Data("GDP1".utf8)
    private static let version: UInt32 = 1
    private static let fieldCount: UInt16 = 5

    public init(baseURL: URL, accountID: String, sessionID: String, pairingSecret: Data) throws {
        guard pairingSecret.count == 32 else { throw PayloadError.malformed }
        guard baseURL.scheme == "https", baseURL.host != nil,
              baseURL.query == nil, baseURL.fragment == nil else {
            throw PayloadError.malformed
        }
        guard Self.isCanonicalUUID(accountID), Self.isCanonicalUUID(sessionID) else {
            throw PayloadError.malformed
        }
        self.baseURL = baseURL
        self.accountID = accountID
        self.sessionID = sessionID
        self.pairingSecret = pairingSecret
    }

    /// Canonical bytes. The field set is fixed and ascending, so this is a
    /// function of the values alone — which is what makes the decoder's
    /// re-encode check meaningful.
    public func encoded() -> Data {
        var out = Self.magic
        out.append(contentsOf: Self.fieldCount.bigEndianBytes)
        Self.appendField(&out, id: 1, value: Data(Self.version.bigEndianBytes))
        Self.appendField(&out, id: 2, value: Data(baseURL.absoluteString.utf8))
        Self.appendField(&out, id: 3, value: Data(accountID.utf8))
        Self.appendField(&out, id: 4, value: Data(sessionID.utf8))
        Self.appendField(&out, id: 5, value: pairingSecret)
        return out
    }

    /// Base64URL without padding, which is what a QR actually carries.
    public func encodedText() -> String {
        encoded().base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    public static func decode(text: String) throws -> SyncPairingPayload {
        guard let bytes = base64URLDecode(text) else { throw PayloadError.malformed }
        let payload = try decode(bytes)
        // Re-encoding is the whole canonicality check: trailing bytes, a
        // missing field, reordered fields, a lowercase UUID and a padded or
        // otherwise non-canonical Base64 spelling all fail right here.
        guard payload.encoded() == bytes, payload.encodedText() == text else {
            throw PayloadError.notCanonical
        }
        return payload
    }

    public static func decode(_ bytes: Data) throws -> SyncPairingPayload {
        guard bytes.count > 6, bytes.prefix(4) == magic else { throw PayloadError.malformed }
        guard readUInt16(bytes, at: 4) == fieldCount else { throw PayloadError.malformed }

        var offset = 6
        var fields: [UInt16: Data] = [:]
        var order: [UInt16] = []
        while offset < bytes.count {
            guard offset + 7 <= bytes.count else { throw PayloadError.malformed }
            let id = readUInt16(bytes, at: offset)
            guard bytes[bytes.startIndex + offset + 2] == 1 else { throw PayloadError.malformed }
            let length = Int(readUInt32(bytes, at: offset + 3))
            offset += 7
            guard length >= 0, offset + length <= bytes.count else { throw PayloadError.malformed }
            guard fields[id] == nil else { throw PayloadError.malformed }
            fields[id] = Data(bytes[(bytes.startIndex + offset)..<(bytes.startIndex + offset + length)])
            order.append(id)
            offset += length
        }
        guard offset == bytes.count, order == [1, 2, 3, 4, 5] else { throw PayloadError.malformed }

        guard let versionField = fields[1], versionField.count == 4 else { throw PayloadError.malformed }
        guard readUInt32(versionField, at: 0) == version else { throw PayloadError.unsupportedVersion }

        guard let urlField = fields[2], let urlText = String(data: urlField, encoding: .utf8),
              let url = URL(string: urlText),
              let accountField = fields[3], let account = String(data: accountField, encoding: .utf8),
              let sessionField = fields[4], let session = String(data: sessionField, encoding: .utf8),
              let secret = fields[5] else {
            throw PayloadError.malformed
        }
        return try SyncPairingPayload(
            baseURL: url,
            accountID: account,
            sessionID: session,
            pairingSecret: secret
        )
    }

    private static func appendField(_ out: inout Data, id: UInt16, value: Data) {
        out.append(contentsOf: id.bigEndianBytes)
        out.append(1)
        out.append(contentsOf: UInt32(value.count).bigEndianBytes)
        out.append(value)
    }

    private static func readUInt16(_ data: Data, at offset: Int) -> UInt16 {
        let base = data.startIndex + offset
        return (UInt16(data[base]) << 8) | UInt16(data[base + 1])
    }

    private static func readUInt32(_ data: Data, at offset: Int) -> UInt32 {
        let base = data.startIndex + offset
        return (0..<4).reduce(UInt32(0)) { ($0 << 8) | UInt32(data[base + $1]) }
    }

    private static func base64URLDecode(_ text: String) -> Data? {
        guard !text.contains("="), !text.contains("+"), !text.contains("/") else { return nil }
        var padded = text.replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        while padded.count % 4 != 0 { padded.append("=") }
        return Data(base64Encoded: padded)
    }

    static func isCanonicalUUID(_ value: String) -> Bool {
        value.range(of: "^[0-9A-F]{8}(-[0-9A-F]{4}){3}-[0-9A-F]{12}$", options: .regularExpression) != nil
    }
}

// File-scoped on purpose: `SyncE2EE` declares its own, and two internal
// spellings of the same member would be ambiguous wherever both are visible.
private extension FixedWidthInteger {
    var bigEndianBytes: [UInt8] { withUnsafeBytes(of: bigEndian) { Array($0) } }
}

File: android/app/src/main/java/com/sapiens/gagaodok/sync/SyncPairingPayload.kt
package com.sapiens.gagaodok.sync

import java.net.URI
import java.util.Base64

/**
 * The thing a joining device scans.
 *
 * It carries the address, the account being joined, the session, and the
 * `pairing_secret` — and nothing else. No device token, no master key, no claim
 * secret, no recovery phrase. Whoever holds this payload can attempt to join,
 * so it is shown on screen and never written to a URL, the clipboard, or a log:
 * a URL would make a scanner open a browser, leaving the secret in a history
 * somewhere.
 */
class SyncPairingPayload private constructor(
    val baseUrl: String,
    val accountId: String,
    val sessionId: String,
    val pairingSecret: ByteArray,
) {
    class PayloadException(val reason: Reason) : Exception() {
        enum class Reason { MALFORMED, NOT_CANONICAL, UNSUPPORTED_VERSION }
    }

    /**
     * Canonical bytes. The field set is fixed and ascending, so this is a
     * function of the values alone — which is what makes the decoder's
     * re-encode check meaningful.
     */
    fun encoded(): ByteArray {
        val fields = listOf(
            1 to uint32(VERSION),
            2 to baseUrl.toByteArray(Charsets.US_ASCII),
            3 to accountId.toByteArray(Charsets.US_ASCII),
            4 to sessionId.toByteArray(Charsets.US_ASCII),
            5 to pairingSecret,
        )
        val out = java.io.ByteArrayOutputStream()
        out.write(MAGIC)
        out.write(uint16(fields.size))
        for ((id, value) in fields) {
            out.write(uint16(id))
            out.write(1)
            out.write(uint32(value.size.toLong()))
            out.write(value)
        }
        return out.toByteArray()
    }

    /** Base64URL without padding, which is what a QR actually carries. */
    fun encodedText(): String = Base64.getUrlEncoder().withoutPadding().encodeToString(encoded())

    override fun equals(other: Any?): Boolean =
        other is SyncPairingPayload && encoded().contentEquals(other.encoded())

    override fun hashCode(): Int = encoded().contentHashCode()

    companion object {
        private val MAGIC = "GDP1".toByteArray(Charsets.US_ASCII)
        private const val VERSION = 1L
        private const val FIELD_COUNT = 5
        private val UUID = Regex("[0-9A-F]{8}(-[0-9A-F]{4}){3}-[0-9A-F]{12}")

        fun create(
            baseUrl: String,
            accountId: String,
            sessionId: String,
            pairingSecret: ByteArray,
        ): SyncPairingPayload {
            if (pairingSecret.size != 32) throw malformed()
            val uri = runCatching { URI(baseUrl) }.getOrNull()
            if (uri?.scheme != "https" || uri.host.isNullOrEmpty() ||
                uri.query != null || uri.fragment != null
            ) {
                throw malformed()
            }
            if (!UUID.matches(accountId) || !UUID.matches(sessionId)) throw malformed()
            return SyncPairingPayload(baseUrl, accountId, sessionId, pairingSecret)
        }

        fun decode(text: String): SyncPairingPayload {
            if (text.contains('=') || text.contains('+') || text.contains('/')) throw malformed()
            val bytes = runCatching { Base64.getUrlDecoder().decode(text) }.getOrNull() ?: throw malformed()
            val payload = decode(bytes)
            // Re-encoding is the whole canonicality check: trailing bytes, a
            // missing field, reordered fields, a lowercase UUID and a padded or
            // otherwise non-canonical Base64 spelling all fail right here.
            if (!payload.encoded().contentEquals(bytes) || payload.encodedText() != text) {
                throw PayloadException(PayloadException.Reason.NOT_CANONICAL)
            }
            return payload
        }

        fun decode(bytes: ByteArray): SyncPairingPayload {
            if (bytes.size <= 6 || !bytes.copyOfRange(0, 4).contentEquals(MAGIC)) throw malformed()
            if (readUInt16(bytes, 4) != FIELD_COUNT) throw malformed()

            var offset = 6
            val fields = LinkedHashMap<Int, ByteArray>()
            while (offset < bytes.size) {
                if (offset + 7 > bytes.size) throw malformed()
                val id = readUInt16(bytes, offset)
                if (bytes[offset + 2].toInt() != 1) throw malformed()
                val length = readUInt32(bytes, offset + 3)
                offset += 7
                if (length < 0 || offset + length > bytes.size) throw malformed()
                if (fields.containsKey(id)) throw malformed()
                fields[id] = bytes.copyOfRange(offset, offset + length)
                offset += length
            }
            if (offset != bytes.size || fields.keys.toList() != listOf(1, 2, 3, 4, 5)) throw malformed()

            val version = fields[1] ?: throw malformed()
            if (version.size != 4) throw malformed()
            if (readUInt32(version, 0).toLong() != VERSION) {
                throw PayloadException(PayloadException.Reason.UNSUPPORTED_VERSION)
            }
            return create(
                baseUrl = String(fields[2]!!, Charsets.US_ASCII),
                accountId = String(fields[3]!!, Charsets.US_ASCII),
                sessionId = String(fields[4]!!, Charsets.US_ASCII),
                pairingSecret = fields[5]!!,
            )
        }

        private fun malformed() = PayloadException(PayloadException.Reason.MALFORMED)

        private fun uint16(value: Int) =
            byteArrayOf(((value shr 8) and 0xff).toByte(), (value and 0xff).toByte())

        private fun uint32(value: Long) = byteArrayOf(
            ((value shr 24) and 0xff).toByte(),
            ((value shr 16) and 0xff).toByte(),
            ((value shr 8) and 0xff).toByte(),
            (value and 0xff).toByte(),
        )

        private fun readUInt16(bytes: ByteArray, offset: Int) =
            ((bytes[offset].toInt() and 0xff) shl 8) or (bytes[offset + 1].toInt() and 0xff)

        private fun readUInt32(bytes: ByteArray, offset: Int): Int {
            var value = 0
            for (index in 0 until 4) value = (value shl 8) or (bytes[offset + index].toInt() and 0xff)
            return value
        }
    }
}
